### PR TITLE
but_last() function to return all elements but the last

### PR DESCRIPTION
**Pull request by sphaso**
Hi!             
sorry to drop this PR out of the blue. We use and love your library at work and recently I found myself in need of a function similar to Haskell's `init`. So I thought I'd spend a little time and try to implement it.
 (CI is red for a timeout, same as master)

### CODE REVIEW

File: src/Functional/Init.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ */
+namespace Functional;
+use Functional\Exceptions\InvalidArgumentException;
+use Traversable;
+
+/**
+ * Returns an array containing the elements of the list without its last element.
+ *
+ * @param Traversable|array $collection
+ * @param callable $callback
```
Comment (lstrojny) [inline src/Functional/Init.php]: this parameter does not exist

File: src/Functional/Init.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * @return mixed
+ */
+function init($collection)
```
Comment (lstrojny) [inline src/Functional/Init.php]: Having very little Haskell exposure for myself, do we think this is a good name for a community mostly having a background in PHP and probably more broadly, scripting languages?

Comment (lstrojny) [inline src/Functional/Init.php]: You can use a return type, since we depend on PHP 7 nowadays

File: src/Functional/Init.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+{
+    InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
+
+    $init = [];
+    foreach ($collection as $index => $element) {
```
Comment (lstrojny) [inline src/Functional/Init.php]: One could replace the loop with `$init = is_array($collection) ? $collection : iterator_to_array($collection)` right?

**Comment by sphaso** [comment]
Hi @lstrojny,
I fixed the things you pointed out, thanks! Regarding the name, if it helps, I could find it called `butLast` in Clojure and `initial` in Lodash and Underscore. I'm not good with names, let me know your pick :)

**Comment by lstrojny** [comment]
Thanks for researching naming alternatives, let’s go with "butLast", it’s the most unterstandable I reckon.

**Comment by sphaso** [comment]
Sounds good! Just changed the name.

**Pull request merged**